tensorflow/core/lib/random/random_distributions.h

```diff
@@ -20,18 +20,60 @@ limitations under the License.
 #include <cmath>
 #include <type_traits>
 
-#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
+#include "tensorflow/core/framework/numeric_types.h"
 #include "tensorflow/core/lib/random/philox_random.h"
 #include "tensorflow/core/lib/random/random_distributions_utils.h"
 #include "tensorflow/core/platform/types.h"
+#include "third_party/eigen3/unsupported/Eigen/CXX11/Tensor"
 
 namespace tensorflow {
 namespace random {
 
 // Helper function to convert a 16-bit integer to a half between [0..1).
 PHILOX_DEVICE_INLINE Eigen::half Uint16ToHalf(uint16 x);
+// Helper function to convert a 16-bit integer to a bfloat16 between [1..2).
+PHILOX_DEVICE_INLINE bfloat16 InternalUint16ToBfloat16(uint16 x);
 // Helper function to convert a 16-bit integer to a bfloat16 between [0..1).
-PHILOX_DEVICE_INLINE bfloat16 Uint16ToGfloat16(uint16 x);
+PHILOX_DEVICE_INLINE bfloat16 Uint16ToBfloat16(uint16 x);
+
+// Helper function to format distribution result in vectorization path,
+// it creates Eigen::Tensor and reuses packet feature with SIMD.
+// This function can only work on CPU
+template <class Distribution, class Generator>
+typename Distribution::ResultType VectorizedFormat(
+    Generator* gen, typename Distribution::FormatFunc functor) {
+  typename Generator::ResultType sample;
+  typename Distribution::ResultType result;
+  const int kResultElementCount = Distribution::kResultElementCount;
+  const int inner_count = Generator::kResultElementCount;
+  const int outer_count = kResultElementCount / inner_count;
+  int offset = 0;
+
+  for (int k = 0; k < outer_count; k++) {
+    sample = (*gen)();
+    for (int i = 0; i < inner_count; i++, offset++) {
+      result[offset] = (*functor)(sample[i]);
+    }
+  }
+  // Tail processing if any.
+  // Put the tail condition out of above loop to improve performance:
+  // it will be executed only once and save time on CPU.
+  if (offset < kResultElementCount) {
+    sample = (*gen)();
+    for (int i = 0; offset < kResultElementCount; i++, offset++) {
+      result[offset] = (*functor)(sample[i]);
+    }
+  }
+
+  typedef Eigen::TensorMap<
+      Eigen::Tensor<typename Distribution::ResultElementType, 1,
+                    Eigen::RowMajor, Eigen::DenseIndex>,
+      Eigen::Aligned>
+      Tensor;
+  auto tensor_result = Tensor(&result[0], kResultElementCount);
+  tensor_result = tensor_result - typename Distribution::ResultElementType(1.0);
```

Comment:
Is it only this subtraction that is being vectorized?

Reply:
Yes, the key optimization is to vectorize the minus operation after generator. It's because BF16 scalar minus is very slow.

Reply:
This seems pretty complicated and intrusive just to vectorize a subtraction.

Can't you vectorize the entire UintXXToXX functions?  Set
```
UniformDistribution::kResultElementCount = GreatestCommonMultiple(Generator::kResultElementCount, Eigen::packet_traits<Scalar>::size);
```
and loop over multiples of `Eigen::packet_traits<Scalar>::size` when doing the conversion?

Reply:
The question is, the entire `UintXXToXX` function only generate 4 elements each time, but the max vectorization is 16 on CPU, so:
1. Efficiency is low if only vectorize `UintXXToXX`. BTW, Eigen vectorization path can only be triggered by explicit declaring, and can't be generated automatically by loop unroll.
2. Launch `UintXXToXX` 4 times and get 16 elements to do vectorization to get better performance, that's what I have done

Reply:
Maybe I can add a new vectorization `UintXXToXX` function to handle 16 elements together after launched generator 4 times. I can't remove the scalar `UintXXToXX` since sometime it has to process the remainder.

Reply:
It _may_ cause more overhead, but I doubt it will be significant, since we are dealing with small array sizes and they will all fit in cache.  The values from the generator need to be copied _anyways_, since they come out as `uint32_t`, but we truncate them to `uint16_t` when calling `Uint16ToBfloat16`.  So you may as well have an `InlinedVector<uint16_t, 16>` to store the results of the generator, then pass this to (or load packets from this and pass to) the vectorized `Uint16ToBfloat16`.

`VectorizedUint16ToBfloat16` could work directly with Eigen packets, rather than vectors of 16 (so on Intel, would be elements of 8 bfloat16s for AVX).

Reply:
Make sense, I will have a try, thanks.